types/mdx/types.d.ts

```diff
@@ -1,12 +1,28 @@
 // Internal helper types
 
-// tslint:disable-next-line: strict-export-declare-modifiers
-type FunctionComponent<Props> = (props: Props) => JSX.Element | null;
-// tslint:disable-next-line: strict-export-declare-modifiers
-type ClassComponent<Props> = new (props: Props) => JSX.ElementClass;
-// tslint:disable-next-line: strict-export-declare-modifiers
-type Component<Props> = FunctionComponent<Props> | ClassComponent<Props> | keyof JSX.IntrinsicElements;
-// tslint:disable-next-line: strict-export-declare-modifiers
+// @ts-ignore
+type ElementType = any extends JSX.ElementType ? never : JSX.ElementType;
+
+type FunctionElementType = Extract<ElementType, (props: Record<string, any>) => any>;
+
+type ClassElementType = Extract<ElementType, new (props: Record<string, any>) => any>;
+
+type StringComponent = Extract<keyof JSX.IntrinsicElements, ElementType extends never ? string : ElementType>;
+
+type FunctionComponent<Props> = ElementType extends never
+    ? (props: Props) => JSX.Element | null
+    : FunctionElementType extends never
+    ? never
+    : (props: Props) => ReturnType<FunctionElementType>;
+
+type ClassComponent<Props> = ElementType extends never
+    ? new (props: Props) => JSX.ElementClass
+    : ClassElementType extends never
+    ? never
+    : new (props: Props) => InstanceType<ClassElementType>;
+
+type Component<Props> = FunctionComponent<Props> | ClassComponent<Props> | StringComponent;
```

Comment:
a) maybe good to explain what these types are for. They’re pretty complex.
b) should there be a “RestComponent”, for `JSX.ElementType` that is not a function, class, or `keyof JSX.IntrinsicElements`, that are added here?

Reply:
a) Yeah, good point. I feel more confident making further updates once there is a solution for CI failing on the `@ts-ignore`.
b) `RestComponent` (e.g. unknown strings / other primitives) would still not be valid JSX components accordng to TypeScript: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEA7KBbEBnADlMCBSAygBrwDeAUPFfAC4CemCAohCGojQCoMIC8SAVxQAjEDHgAfeOhowAlogDm5StQU0xAMxwIAkh3mJ0csCzYgO6MqurUZhxQC4yAXxu34AbURDRMZz4iYgC6zqRutm5u5KCQsAhgAPZGNPAASiA4NM5QiHTkSSnwBLLw-ADk9gqK5QXJMsWJaCUOZfDlBE0IVUq1hQ0AckJtAIwq-an2bQA8LfAA9AB8dUXoXS3VM53NsptLKw2BM0MoC8tAA

Reply:
Re b), that seems like a TS bug, [see also this example](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEA7KBbEBnADlMCBSAygBrwDeAUPFfAC4CemCAohCGojQCoMIC886GjACWiAOblK1UTRAwAZjgQBJDiMTphYFmxAd0ZAL7lj5UJFgIwAew014AJRA4aALnhREdSTbvwAgvD8AOQARsE+toIeQfAAPIEA9AB8QA)?

```ts
declare namespace JSX {
    type ElementType = string

    interface IntrinsicElements {}
}

declare const React: any

const A = 'b'

const a = <A />
// > Property 'b' does not exist on type 'JSX.IntrinsicElements'.(2339)
// > JSX element type 'A' does not have any construct or call signatures.(2604)
```

That is: `ElementType` allows any string, `IntrinsicElements` does not particularly include this string. Practical example of this is custom elements: `<a-b>`?
Not sure if we can even do that here, because the example above is without any MDX types.

---

Could we replace `StringComponent` with a `RestComponent`, such as:

```js
type RestComponent = Exclude<
  ElementType,
  ClassElementType | FunctionElementType
>;
```

Which currently works exactly the same, but feels safer to me, as it makes sure we don’t drop stuff accidentally, and would allow “unknown strings / other primitives” (/other non-promitives if ever they are added?) when TS starts supporting those?

Reply:
I don’t think this is a bug. I’m not aware of any plans to change how this works. The goal is to make it work with the way TypeScript currently works, not how you or I think it should work.

Reply:
> TypeScript currently works

How TS works in the future is important too?

It’s been a month so I lost the context.

But a `RestComponent` seems more future proof?

Reply:
This is not a bug. The goal is to allow components that are valid JSX elements. This is achieved by the current code. Your suggestion to use `RestComponent` yields types that are not valid JSX components.

I am not aware of plans to change how this works in TypeScript. The current solution is future proof.

With the current implementation it is possible to use custom elements. It works the same as using it in JSX/React. The user needs to register it by augmenting `JSX.IntrinsicElements`. This makes it available in both JSX and MDX components.

Reply:
I’m marking this as resolved, as I have added comments and the types work as expected.